mlx_graphs/nn/conv/rel_conv.py

```diff
@@ -0,0 +1,355 @@
+    The input shape is expected to be [batch_size, num_nodes, input_dim]
+
+    Alternatively, the layer can work as a standard relational conv
+    with shapes [num_nodes, input_dim].
+
+    Note that this implementation materializes all edge messages and is O(E).
+    The complexity can be further reduced by adopting the O(V) `rspmm` C++ kernel
+    from the NBFNet-PyG repo to the MLX framework (not implemented here).
+
+    Args:
+        input_dim: input feature dimension (same for node and edge features)
+        output_dim: output node feature dimension
+        num_relation: number of unique relations in the graph
+        message_func: "transe" (sum), "distmult" (mult),
+            "rotate" (complex rotation). Default: ``distmult``
+        aggregate_func: "add", "mean", or "pna". Default: ``add``
+        layer_norm: whether to use layer norm
+            (often crucial to the performance). Default: ``True``
+        activation: non-linearity. Default: ``relu``
+        dependent: whether to use separate relation embedding matrix
+             ``False`` or build relations from the input relations ``True``
+        node_dim: for 3D batches, specified which dimension contains all nodes.
+            Default: ``0``
+
+    Example:
+
+    .. code-block:: python
+
+        import mlx.core as mx
+        import mlx.nn as nn
+        from mlx_graphs.nn import GeneralizedRelationalConv
+
+        input_dim = 16
+        output_dim = 16
+        num_relations = 3
+
+        conv = GeneralizedRelationalConv(input_dim, output_dim, num_relations)
+
+        batch_size = 2
+        edge_index = mx.array([[0, 1, 2, 3, 4], [0, 0, 1, 1, 3]])
+        edge_types = mx.array([0, 0, 1, 1, 2])
+        boundary = mx.random.uniform(0, 1, shape=(batch_size, 5, 16)
+        size = (node_features.shape[0], node_features.shape[0])
+
+        layer_input = boundary  # initial node features which will be updated
+        h = conv(edge_index, layer_input, edge_type, boundary, size=size)
+
+        # optional: residual connection if input dim == output dim
+        h = h + layer_input
+        layer_input = h
+
+        # another conv type where relations are obtained from the additional
+        # query tensor
+        query = mx.random.uniform(0, 1, shape=(batch_sim, 16))
+        conv2 = GeneralizedRelationalConv(
+            input_dim, output_dim, num_relations, dependent=True)
+
+        h = conv2(edge_index, layer_input, edge_type, boundary, query, size=size)
+
+    """
+
+    eps = 1e-6
+
+    def __init__(
+        self,
+        in_features_dim: int,
+        out_features_dim: int,
+        num_relations: int,
+        message_func: MessageFunctions = "distmult",
+        aggregate_func: AggegationFunctions = "add",
+        layer_norm: bool = True,
+        activation: str = "relu",
+        dependent: bool = False,
+        node_dim: int = 0,
+        **kwargs,
+    ):
+        kwargs.setdefault("aggr", "add")
```

Comment:
is this necessary? `MessagePassing` has `aggr` set to `"add"` by default

Reply:
Removed in https://github.com/mlx-graphs/mlx-graphs/pull/111/commits/a694a2574d5c96991c6d6d776494b9e5bbe252a3